flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogStreamHandleReaderWithCache.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final String CACHE_FILE_SUB_DIR = "dstl-cache-file";
+    private static final String CACHE_FILE_PREFIX = "dstl";
+
+    // reference count == 1 means only cache component reference the cache file
+    private static final int NO_USING_REF_COUNT = 1;
+
+    private final File[] cacheDirectories;
+    private final AtomicInteger next;
+
+    private final ConcurrentHashMap<Path, RefCountedFile> cache = new ConcurrentHashMap<>();
+    private final ScheduledExecutorService cacheCleanScheduler;
+    private final long cacheIdleMillis;
+
+    ChangelogStreamHandleReaderWithCache(Configuration config) {
+        this.cacheDirectories =
+                Arrays.stream(ConfigurationUtils.parseTempDirectories(config))
+                        .map(path -> new File(path, CACHE_FILE_SUB_DIR))
+                        .toArray(File[]::new);
+        this.next = new AtomicInteger(new Random().nextInt(this.cacheDirectories.length));
+
+        this.cacheCleanScheduler =
+                SchedulerFactory.create(1, "ChangelogCacheFileCleanScheduler", LOG);
+        this.cacheIdleMillis = config.get(CACHE_IDLE_TIMEOUT).toMillis();
+    }
+
+    @Override
+    public DataInputStream openAndSeek(StreamStateHandle handle, Long offset) throws IOException {
+        if (!canBeCached(handle)) {
+            return wrapAndSeek(handle.openInputStream(), offset);
+        }
+
+        final FileStateHandle fileHandle = (FileStateHandle) handle;
+        final RefCountedFile refCountedFile = getRefCountedFile(fileHandle);
+
+        FileInputStream fin = openAndSeek(refCountedFile, offset);
+
+        return wrapStream(fileHandle.getFilePath(), fin);
+    }
+
+    private boolean canBeCached(StreamStateHandle handle) throws IOException {
+        if (handle instanceof FileStateHandle) {
+            FileStateHandle fileHandle = (FileStateHandle) handle;
+            return fileHandle.getFilePath().getFileSystem().isDistributedFS();
+        } else {
+            return false;
+        }
+    }
+
+    private RefCountedFile getRefCountedFile(FileStateHandle fileHandle) {
+        return cache.compute(
+                fileHandle.getFilePath(),
+                (key, oldValue) -> {
+                    if (oldValue == null) {
+                        oldValue = downloadToCacheFile(fileHandle);
+                    }
+                    oldValue.retain();
+                    return oldValue;
+                });
+    }
+
+    private RefCountedFile downloadToCacheFile(FileStateHandle fileHandle) {
+        RefCountedFile refCountedFile = null;
+
+        try (FSDataInputStream inputStream = fileHandle.openInputStream()) {
+            File directory = cacheDirectories[next.getAndIncrement() % cacheDirectories.length];
+            File file = File.createTempFile(CACHE_FILE_PREFIX, null, directory);
+
+            IOUtils.copyBytes(wrap(inputStream), new FileOutputStream(file));
```

Comment:
1. No need to `wrap` here?
2. The input stream is closed twice; I think we should use an overloaded `copyBytes` with `close=false`

Reply:
Hi @rkhachatryan , thank you for your patience !  I‘m continuing to update this pr as you suggested.

1. As mentioned in the [design document](https://docs.google.com/document/d/1uTYNutW52cz8GUxzTfIJojctEWD_orucYXs2MU6oHOo/#heading=h.4n9l6huj81wl), the decompressed content is stored in the cache file. This eliminates the need for repeated decompression each time read, and you can seek directly to the desired position.
2. you are right , how about move `inputStream` into try body ?

Reply:
Hi @zoltar9264 , no worries, thanks for updating the PR! :)

1. Makes sense
2. Both input and output streams should be closed once eventually; I think it can be done better using try-with-resources, rather than inside `copyBytes`. I don't have a strong preference though.